devito/archinfo.py

```diff
@@ -117,6 +119,121 @@ def get_cpu_brand():
     return cpu_info
 
 
+@memoized_func
+def get_gpu_info():
+    """Attempt GPU info autodetection."""
+
+    # Filters out virtual GPUs from a list of GPU dictionaries
+    def filter_real_gpus(gpus):
+        def is_real_gpu(gpu):
+            return 'virtual' not in gpu['product'].lower()
+        return filter(is_real_gpu, gpus)
+
+    # The following functions of the form cmd_gpu_info(...) attempt obtaining GPU
+    #   information using 'cmd'
+    # The currently supported ways of obtaining GPU information (in order of attempt) are:
+    #   - 'lshw' from the command line
+    #   - 'lspci' from the command line
+
+    def lshw_gpu_info(text):
+        def lshw_single_gpu_info(text):
+            # Separate the output into lines for processing
+            lines = text.replace('\\n', '\n')
+            lines = lines.splitlines()
+
+            # Define the processing functions
+            if lines:
+                def extract_gpu_info(keyword):
+                    for line in lines:
+                        if line.lstrip().startswith(keyword):
+                            return line.split(':')[1].lstrip()
+
+                def parse_product_arch():
+                    for line in lines:
+                        if line.lstrip().startswith('product') and '[' in line:
+                            return line.split('[')[1].split(']')[0]
+                    return 'unspecified'
+
+                # Populate the information
+                gpu_info = {}
+                gpu_info['product'] = extract_gpu_info('product')
+                gpu_info['architecture'] = parse_product_arch()
+                gpu_info['vendor'] = extract_gpu_info('vendor')
+                gpu_info['physicalid'] = extract_gpu_info('physical id')
+
+                return gpu_info
+
+        # Parse the information for all the devices listed with lshw
+        devices = text.split('display')[1:]
+        gpu_infos = [lshw_single_gpu_info(device) for device in devices]
+        return filter_real_gpus(gpu_infos)
+
+    def lspci_gpu_info(text):
+        # Note: due to the single line descriptive format of lspci, 'vendor'
+        #   and 'physicalid' elements cannot be reliably extracted so are left None
+
+        # Separate the output into lines for processing
+        lines = text.replace('\\n', '\n')
+        lines = lines.splitlines()
+
+        gpu_infos = []
+        for line in lines:
+            # Graphics cards are listed as VGA or 3D controllers in lspci
+            if 'VGA' in line or '3D' in line:
+                gpu_info = {}
+                gpu_info['product'] = line.split(':')[1].split(':')[1].split(':').lstrip()
+
+                if '[' in line:
+                    gpu_info['architecture'] = line.split('[')[1].split(']')[0]
```

Comment:
I wonder whether here, and actually in a few other places in this funcion, you should actually use a regex